File: workspace/2023-03-12/mathematics.ruby/mathematics.rb

def count(xs)
    xs.reduce(0, &lambda{|count, x| count + 1})
end

def minimum(xs)
    less = lambda{|a, b| a < b ? a : b}
    xs.reduce(xs[0], &less)
end

def maximum(xs)
    greater = lambda{|a, b| a > b ? a : b}
    xs.reduce(xs[0], &greater)
end

def sum(xs)
    xs.reduce(0, :+)
end

def product(xs)
    xs.reduce(1, :*)
end

def average(xs)
    sum(xs).to_f / count(xs)
end

def variance(xs)
    sum(xs.map{|x| (x - average(xs)) ** 2}) / count(xs)
end

def stddev(xs)
    Math.sqrt(variance(xs))
end

File: workspace/2023-07-10/mathematics.python/mathematics.py
from functools import reduce
from operator import add


def count(xs):
    return reduce(lambda n, x: n + 1, xs, 0)


def sum(xs):
    return reduce(add, xs, 0)


def average(xs):
    return sum(xs) / count(xs)


def minimum(xs):
    def less(a, b):
        if a < b:
            return a
        else:
            return b
    return reduce(less, xs[1:], xs[0])


def maximum(xs):
    def greater(a, b):
        if a > b:
            return a
        else:
            return b
    return reduce(greater, xs[1:], xs[0])


def variance(xs):
    return sum([(x - average(xs)) ** 2 for x in xs]) / count(xs)


def stddev(xs):
    return variance(xs) ** 0.5

